### PR TITLE
Fix all() unroll for non-generators/non-list comprehensions

### LINKED ISSUE #5358: pytest 4.6.0 breaks matplotlib test gathering with `TypeError: cannot unpack non-iterable NoneType object`
This build: https://travis-ci.org/matplotlib/matplotlib/jobs/540086010 passed fine, with pytest 4.5.0. The next build from the same code, initiated via cron and using pytest 4.6.0: https://travis-ci.org/matplotlib/matplotlib/jobs/540212001 failed with

```
Calling pytest with the following arguments: -raR --maxfail=50 --timeout=300 --durations=25 --cov-report= --cov=lib -n 2 --log-level=DEBUG
============================= test session starts ==============================
platform linux -- Python 3.7.1, pytest-4.6.0, py-1.7.0, pluggy-0.12.0
rootdir: /home/travis/build/matplotlib/matplotlib, inifile: pytest.ini, testpaths: lib
plugins: forked-1.0.2, cov-2.7.1, rerunfailures-7.0, xdist-1.28.0, timeout-1.3.3, faulthandler-1.6.0
timeout: 300.0s
timeout method: signal
timeout func_only: False
gw0 I / gw1 I
gw0 [5784] / gw1 [5784]
...
_______ ERROR collecting lib/matplotlib/tests/test_constrainedlayout.py ________
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:963: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:902: in _find_spec
    ???
<frozen importlib._bootstrap>:876: in _find_spec_legacy
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/opt/python/3.7.1/lib/python3.7/ast.py:262: in visit
    return visitor(node)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
E   TypeError: cannot unpack non-iterable NoneType object
_______ ERROR collecting lib/matplotlib/tests/test_constrainedlayout.py ________
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:963: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:902: in _find_spec
    ???
<frozen importlib._bootstrap>:876: in _find_spec_legacy
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/opt/python/3.7.1/lib/python3.7/ast.py:262: in visit
    return visitor(node)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
E   TypeError: cannot unpack non-iterable NoneType object
____________ ERROR collecting lib/matplotlib/tests/test_mathtext.py ____________
In test_mathtext_rendering: 1 parameter sets specified, with different number of ids: 83
```

The operating system is Ubuntu 16.04.6 LTS (xenial).

I don't know how to unpack the line numbers for the breaking tests from the above, but [test_constrainedlayout.py](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_constrainedlayout.py) has the following asserts:

```
    assert all(ax.get_position().extents == ax2.get_position().extents)
    assert all(ax.get_position().extents == ax2.get_position().extents)
    np.testing.assert_allclose(extents0, extents1)
    np.testing.assert_allclose(extents0, extents1)
```

In addition, there are multiple tests using matplotlib's `@image_comparison` decorator, but there are so many of those all over the codebase that I think then we'd see much more failures.

- [x] a detailed description of the bug or suggestion
- [ ] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [ ] minimal example if possible

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #5358

### CODE REVIEW

File: src/_pytest/assertion/rewrite.py [hunk 1]
```diff
@@ -954,7 +954,9 @@ def visit_Call_35(self, call):
         visit `ast.Call` nodes on Python3.5 and after
         """
         if isinstance(call.func, ast.Name) and call.func.id == "all":
```
Comment (asottile) [inline src/_pytest/assertion/rewrite.py]: I think we should put the guard here instead (functions returning `Optional[T]` are a bit odd)

there's also another edgecase which is not handled:

```python
if all(*f):
    ...
```

(in python2 this will `IndexError`)

Comment (asottile) [reply src/_pytest/assertion/rewrite.py]: Here is how I handle this in `pyupgrade`: https://github.com/asottile/pyupgrade/blob/8ba7b8b1844f76d2f3a8233cb1a632f3358fea9f/pyupgrade.py#L1368-L1378

though I think in this case it's not super necessary to identify it as starargs or not (since in py2 you'll end up with `args=[]` and in py35+ you'll end up with `args=Starred(...)` both of which would be caught by a length and isinstance check

So basically:

```python
if (
        isinstance(call.func, ast.Name) and
        call.func.id == 'all' and
        len(call.args) == 1 and
        isinstance(call.args[0], (...))
):
    return self._visit_call(...)
```

Might even be useful to isolate to a function so you can do

```python
if _is_all_call(call):
    return self._visit_call(call)
```

Comment (nicoddemus) [reply src/_pytest/assertion/rewrite.py]: > I think we should put the guard here instead (functions returning Optional[T] are a bit odd)

Good point, thanks. Done.

> if all(*f):

What do you mean? The assertion rewriter only kicks in for `assert` statements, not for `if`s.

Comment (asottile) [reply src/_pytest/assertion/rewrite.py]: ah yeah I meant `assert all(*f)`:

```python
def test():
    f = ((1,))
    assert all(*f)
```

```console
$ pytest t.py 
============================= test session starts ==============================
platform linux2 -- Python 2.7.15rc1, pytest-4.6.0, py-1.8.0, pluggy-0.12.0
rootdir: /tmp
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
____________________________ ERROR collecting t.py _____________________________
venv/local/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/usr/lib/python2.7/ast.py:241: in visit
    return visitor(node)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
/usr/lib/python2.7/ast.py:241: in visit
    return visitor(node)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:1013: in visit_Call_legacy
    return self._visit_all(call)
venv/local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:983: in _visit_all
    if not isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp)):
E   IndexError: list index out of range
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.36 seconds ============================
```

File: src/_pytest/assertion/rewrite.py [hunk 2]
```diff
@@ -949,11 +949,19 @@ def visit_BinOp(self, binop):
         res = self.assign(ast.BinOp(left_expr, binop.op, right_expr))
         return res, explanation
 
+    def _is_any_call_with_generator_or_list_comprehension(self, call):
+        """Return True if the Call node is an 'any' call with a generator or list comprehension"""
+        return (
+            isinstance(call.func, ast.Name)
+            and call.func.id == "all"
+            and isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp))
```
Comment (asottile) [inline src/_pytest/assertion/rewrite.py]: this still needs a length check -- can you add this as a test?

```python
def test():
    f = ((1,))
    assert all(*f)
```

File: src/_pytest/assertion/rewrite.py [hunk 3]
```diff
@@ -949,11 +949,19 @@ def visit_BinOp(self, binop):
         res = self.assign(ast.BinOp(left_expr, binop.op, right_expr))
         return res, explanation
 
+    def _is_any_call_with_generator_or_list_comprehension(self, call):
```
Comment (asottile) [inline src/_pytest/assertion/rewrite.py]: can be `@staticmethod` (does not use `self`)

File: src/_pytest/assertion/rewrite.py (hunk 1 above)
Comment (nicoddemus) [reply src/_pytest/assertion/rewrite.py]: Done! 👍

File: src/_pytest/assertion/rewrite.py (hunk 3 above)
Comment (nicoddemus) [reply src/_pytest/assertion/rewrite.py]: Done! 👍

File: src/_pytest/assertion/rewrite.py (hunk 2 above)
Comment (nicoddemus) [reply src/_pytest/assertion/rewrite.py]: Done! 👍

**Comment by asottile** [comment]
I can backport this one, I want to try a different procedure and write some stuff up so it's less mysterious

**Comment by nicoddemus** [comment]
> I can backport this one

Sure, please go ahead! 👍 😁

**Comment by nicoddemus** [comment]
btw I think we need to make a 4.6.1 release with this fix, as it simply breaks the entire test run and without a clear workaround other than disabling assertion rewriting entirely.

**Comment by asottile** [comment]
happy to do that as well 👍 -- shall I release from master or should we try our first release from the maintenance branch?

**Comment by nicoddemus** [comment]
> shall I release from master or should we try our first release from the maintenance branch?

Let's do it from the maintenance branch. 👍
